Решение по 16 спринту

iliaalii opened the pull request:
- добавлены запросы на предметы
- приложение разделено на многомодульное [server, gateway]
- добавлены тесты

iliaalii committed:
Решение по 16 спринту:
 - добавлены запросы на предметы

iliaalii committed:
Решение по 16 спринту:
 - приложение разделено на многомодульное [server, gateway]
 - добавлены тесты

iliaalii committed:
исправление по 16 спринту:
 - нужно больше тестов

iliaalii committed:
исправление по 16 спринту:
 - не проходит тесты...

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/gateway/exception/ErrorHandler.java:
```diff
@@ -45,13 +35,6 @@ public ErrorResponse handleConstraintViolation(final ConstraintViolationExceptio
-    }
-
     @ExceptionHandler
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
```
Вот тут у меня  вопрос. Как известно: 

Ошибка  500 Internal Server Error - это внутренняя серверная проблема, ее возникновение обусловлено тем, что от клиента (браузера, десктопной программы и т. п.) в сторону сервера поступает запрос, а сервер не может корректно его обработать. 

Мы же еще находимся на гейтвей, так? Проверяем данные, нашли, что что-то не так с входными параметрами... Причём тут сервер? 

Список  возможных статусов, в принципе, известен заранее...

iliaalii replied:
логично, почистил.
не все пути просчитал...

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/gateway/item/controller/ItemController.java:
```diff
@@ -0,0 +1,62 @@
+    @GetMapping("/search")
+    @ResponseStatus(HttpStatus.OK)
+    public ResponseEntity<Object> search(@RequestHeader(HEADER_USER_ID) @Positive long userId, @RequestParam String text) {
+        return itemClient.search(userId, text);
```
Лучше было бы прямо здесь проверить входной параметр text. И, если текст запроса пустой, или недопустимый, например, состоит из пробелов, то прямо отсюда и вернуть результат, не обращаться к серверу. Вернуть можно, например, пустую коллекцию.

iliaalii replied:
перенес

avfyodorov reviewed (commented):
Добрый день, Илья! 

Очень хорошая работа, от меня всего пара уточнений.

iliaalii committed:
исправление по 16 спринту:
 - возврат пустого списка при некорректном запросе поиска вещи по ключевому слову, перенесены в gateway
 - из обработчика ошибок gateway, убрано обработка ошибок сервера (INTERNAL_SERVER_ERROR)

avfyodorov reviewed (commented):
Добрый день, Илья!

Замечаний нет.
Работа принята.
